### PR TITLE
Support Slurm Autorequeue for Array Jobs

### LINKED ISSUE #15022: Slurm Autorequeue Doesn't Work for Job Arrays
### Bug description

It looks like the slurm auto-requeue doesn't work right when a job is part of a job array. I started noticing weird bugs where my jobs would ask to requeue at "weird" times or they would "fail to requeue" but then mysteriously restart or they wouldn't requeue at all. The common thread is that I was using job arrays to manage a batch of related experiments automatically. 

It looks like when this happens you need to make sure you tell `scontrol` to requeue the specific array job which can be found with the environment variables `SLURM_ARRAY_JOB_ID` and `SLURM_ARRAY_TASK_ID` so you would do something like `scontrol requeue $SLURM_ARRAY_JOB_ID_$SLURM_ARRAY_TASK_ID`. What lightning was doing was requeueing the entire job which was trying to get both of the jobs in the array back on the queue and causing weird issues.

Fixing this seems to be really simple, just detect if those array variables are set and make a different call to `scontrol`. I will PR this ASAP because it's suddenly become mission critical for me (long story), what is your deadline for getting a feature into 1.8?

ref
https://github.com/facebookincubator/submitit/blob/main/submitit/slurm/slurm.py#L187
https://github.com/facebookincubator/submitit/blob/4cf1462d7216f9dcc530daeb703ce07c37cf9d72/submitit/slurm/slurm.py#L192
https://github.com/facebookincubator/submitit/blob/4cf1462d7216f9dcc530daeb703ce07c37cf9d72/submitit/core/job_environment.py#L82
https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/trainer/connectors/signal_connector.py#L82

### How to reproduce the bug

```python
* Try to run an array job using lightning and with autorequeue enabled
* Wait a while, if you're unlucky weird stuff will happen
```

### Error messages and logs

_No response_

### Environment

```

#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): 
#- PyTorch Lightning Version (e.g., 1.5.0): 1.8.0.dev0 (built myself from master)
#- Lightning App Version (e.g., 0.5.2): 
#- PyTorch Version (e.g., 1.10): 1.12.1+cu116
#- Python version (e.g., 3.9): 3.8.10
#- OS (e.g., Linux): Linux
#- CUDA/cuDNN version: 
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source): 
#- Running environment of LightningApp (e.g. local, cloud): 

```

### More info

_No response_

### PR DESCRIPTION
**Pull request by Queuecumber**
## What does this PR do?

This PR adds support for autorequeueing slurm array jobs. 

Array jobs are groups of jobs which are grouped in some meaningful way. Slurm provides facilities to quickly (i.e. with less load on the scheduler) create such jobs. 

These jobs behave slightly differently from regular jobs. 

Although they have `SLURM_JOB_ID` set, and it is unique for each job in the array, some commands, notably `scontrol`, differ in their meaning when the `JOB_ID` is used vs. the array specific `SLURM_ARRAY_JOB_ID` and `SLURM_ARRAY_TASK_ID`. For example, if `scontrol` is called with the `SLURM_JOB_ID` of the "parent job" it will apply the `scontrol` command to all array jobs. In order to requeue (for example) only a specific array job, the string `<array job id>_<array_task_id>` needs to be used.

For example if a job has `JOB_ID` 1234, the first array job will have `ARRAY_JOB_ID` 1234 and `TASK_ID` 1 and the second job will have `JOB_ID` 1235 `ARRAY_JOB_ID` 1234 and `TASK_ID` 2. In this case  `scontrol requeue 1234`  would requeue both jobs in the array. `scontrol requeue 1234_1` would requeue only the first job in the array and `scontrol requeue 1234_2` would requeue only the second.

Since lightning was only checking the `JOB_ID` this was causing it to prematurely requeue some array jobs. This was causing lots of weird issues like failed requeues and timeouts with no requeue attempted. 

Fixes #15022

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Queuecumber** [comment]
I'm pretty confident in the test cases I wrote but I'm leaving this in draft mode until I get a chance to run it on one of our clusters. Should be before next week.

**Comment by Queuecumber** [comment]
This is confirmed working on one of our clusters so I think we're good to go

**Comment by Queuecumber** [comment]
Not sure why the python 3.7 tests are failing, when I run them locally with python 3.7.14 they work

**Review by awaelchli — APPROVED** [review]
@Queuecumber I fixed the tests. 
Thanks for extending this feature!!

File: tests/tests_pytorch/trainer/connectors/test_signal_connector.py [hunk 1]
```diff
@@ -100,6 +100,32 @@ def test_auto_requeue_custom_signal_flag(auto_requeue, requeue_signal):
...
+@RunIf(skip_windows=True)
+@mock.patch("pytorch_lightning.trainer.connectors.signal_connector.call")
+@mock.patch("pytorch_lightning.trainer.Trainer.save_checkpoint", mock.MagicMock())
+@mock.patch.dict(os.environ, {"SLURM_JOB_ID": "12346", "SLURM_ARRAY_JOB_ID": "12345", "SLURM_ARRAY_TASK_ID": "2"})
+def test_auto_requeue_array_job(call_mock):
+    call_mock.return_value = 0
+    trainer = Trainer(plugins=[SLURMEnvironment()])
+    connector = SignalConnector(trainer)
+    connector.slurm_sigusr_handler_fn(None, None)
+    call_mock.assert_called_once_with(["scontrol", "requeue", "12345_2"])
```
[the remark this answers is not in the document]
Comment (Queuecumber) [reply tests/tests_pytorch/trainer/connectors/test_signal_connector.py:125]: 👍

File: src/pytorch_lightning/trainer/connectors/signal_connector.py [hunk 2]
```diff
@@ -78,7 +78,13 @@ def slurm_sigusr_handler_fn(self, signum: _SIGNUM, frame: FrameType) -> None:
 
         if self.trainer.is_global_zero:
             # find job id
-            job_id = os.environ["SLURM_JOB_ID"]
+            array_job_id = os.getenv("SLURM_ARRAY_JOB_ID")
```
[the remark this answers is not in the document]
Comment (Queuecumber) [reply src/pytorch_lightning/trainer/connectors/signal_connector.py:81]: Yeah actually the entire function should probably be in the slurm environment since that would keep all the slurm stuff in one place

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/signal_connector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/signal_connector.py:81]: Resolve to merge, but you should follow-up on this

[the remark this answers is not in the document]
Comment (Queuecumber) [reply src/pytorch_lightning/trainer/connectors/signal_connector.py:81]: Thanks for merging, I definitely could have included this in the PR if you wanted, although it's somewhat of an invasive change there's probably not much risk to it either
